From d1e3ca5ac1130c089966b610c82f7d5ba098c321 Mon Sep 17 00:00:00 2001
From: Pavan S <psreenivas@habana.ai>
Date: Tue, 23 Aug 2022 08:12:32 +0300
Subject: [PATCH] [SW-101468] gaudi2: erom: do not deassert wd gpio in automated recovery

The CPLD will deassert GPIO2 when wd gpio(GPIO62) is deasserted
by firmware, In case of automated recovery it is agreed to deassert
the wd gpio only after successful firmware update i,e. in zephyr
agent Hence do not deassert wd gpio in automated recovery.

Signed-off-by: Pavan S <psreenivas@habana.ai>
Change-Id: Ieb14d46a4b2aaddfb401862668ce0d4522096285
---

diff --git a/arch/gaudi2/common/common.c b/arch/gaudi2/common/common.c
index 3f4d253..179a7af 100644
--- a/arch/gaudi2/common/common.c
+++ b/arch/gaudi2/common/common.c
@@ -9,6 +9,7 @@
 
 #include <psoc_global_conf_regs.h>
 #include <psoc_global_conf_masks.h>
+#include <psoc_gpio0_regs.h>
 #include <psoc_gpio1_regs.h>
 #include <psoc_gpio2_regs.h>
 #include <psoc_gpio3_regs.h>
@@ -1014,10 +1015,22 @@
 }
 
 #define WD_ASSERTION_GPIO	14
+#define FORCE_RECOVERY_MASK	BIT(2)
 void preboot_clear_wd_gpio(void)
 {
 	u32 val, mask;
 
+	/*
+	 * Deassert WD GPIO62 only when GPIO2 is not asserted by CPLD
+	 * when GPIO2 is asserted (automated recovery), de-asserting GPIO62
+	 * is done in agent after successful firmware update which indicates
+	 * CPLD to de-assert GPIO2, so that upon next reset arcboot rom tries
+	 * to load the updated ppboot image.
+	 */
+	val = regmap_readl(mmPSOC_GPIO0_DATA) & FORCE_RECOVERY_MASK;
+	if (!val)
+		return;
+
 	mask = BIT(WD_ASSERTION_GPIO);
 	regmap_writel(mask, mmPSOC_GPIO3_OUTENSET);
 	val  = regmap_readl(mmPSOC_GPIO3_DATAOUT);
